### PR TITLE
Reconcile Gene Aggregation Feature with Memory Footprint Overhaul

### PR DESCRIPTION
**Pull request by warrenmcg**
Hi @pimentel and Sleuth team,

I went ahead and made changes to reconcile the gene aggregation feature currently available on the most recent stable release, and the overhaul you've already done to improve the memory footprint of Sleuth. I started from commit [8fba175](https://github.com/warrenmcg/sleuth/commit/8fba1753f0236ece508bdf308fcedd103751318f) from the gene_agg branch and accepted all of the commits through to the current stable release (commit [048f055](https://github.com/warrenmcg/sleuth/commit/048f0551a31c4aee6e59b75c86cab46ae1b3ca3a)). I then added in the two commits from the gene_memory branch (through commit [b0d4731](https://github.com/warrenmcg/sleuth/commit/b0d47318b007a8fe3b416751e292376499e33287)). This was to help facilitate an automatic merge. The rest is my work.

### **Here is a summary of what has changed from that point onward:**
+ I updated the `reads_per_base_transform` function to use data.table throughout, which results in a significant speed boost. This is used to process the bootstrap matrix and the bootstrap TPM counts, and increased the speed several fold (20 seconds each for the matrix and TPM calculations in one of my own datasets, compared to several minutes each per sample using the old code). Also, because of the details of how I did this work, the TPM counts are done before the est. counts summary.
+ I added in code into the for loop within `sleuth_prep` to process the bootstrap data at the gene-level. This takes advantage of using data.table throughout for fast processing (total processing per sample is <1 min per sample with my own full datasets).
+ I also added in code within `sleuth_prep` to update the sleuth object features to gene level data, not just `bs_summary`: `filter_df`, `filter_bool`, `obs_norm`, and `obs_norm_filt` now report gene-level information. `obs_raw` remains untouched, so it still has the transcript-level data.
+ I added a new `transform_function` option to give people the option of using another transformation function (e.g. using log2 instead of ln; using a different offset instead of 0.5). I extended the primitive $<- function and added a `transform_synced` item to the `sleuth_model` object, to make sure that users can't update the transformation function after having run `sleuth_prep` to protect reproducibility when sharing sleuth results.
+ I added tests to make sure gene aggregation works as expected, and then also added tests to quickly look at whether the whole sleuth pipeline is working properly. I computed preprocessed sleuth results at the transcript and gene level, and do tests to compare these precomputed results to newly run results to make sure the current version of sleuth does not change the sleuth results significantly from before. This can be updated in future releases as other features are added, or redesigned. This can also be a starting point for when to increment major releases.

If this work is accepted, the `gene_summary` function will be obsolete.

### **Other features that are not related:**
The features from previous pull requests are included, specifically #71, #94, #95, and #96. If those are all good, you can just focus on this pull request. If they are not good, we can discuss and make changes accordingly.

### Plan:
I plan to move the code within the for loop to its own function, and then set things up so that the bootstrap reading can be done in parallel, so we can add back in the num_cores option. On my machine, this would reduce what previously took almost three hours with the old code (current release) down to 1-2 minutes (new code with parallel options).

Let me know what you think!

### CODE REVIEW

**Comment by warrenmcg** [comment]
Update: proposed parallelization code is finished. added back in the num_cores option from #94.

File: R/bootstrap.R [hunk 1]
```diff
@@ -359,9 +359,143 @@ dcast_bootstrap.kallisto <- function(obj, units, nsamples = NULL) {
...
+    rm(bs_df)
+    # not sure why, but the melt function always returns a factor,
+    # even when setting variable.factor = F, so I coerce target_id
+    tidy_bs$target_id <- as.character(tidy_bs$target_id)
+    # combine the long tidy table with eff_len and aggregation mappings
+    # note that bootstrap number is treated as "sample" here 
+    # for backwards compatibility
+    tidy_bs <- dplyr::select(tidy_bs, target_id, 
+                             est_counts, sample = bootstrap_num)
+    tidy_bs <- merge(data.table::as.data.table(tidy_bs),
```
Comment (pimentel) [inline R/bootstrap.R]: make all explicit by using scope operator

File: R/bootstrap.R [hunk 2]
```diff
@@ -359,9 +359,143 @@ dcast_bootstrap.kallisto <- function(obj, units, nsamples = NULL) {
...
+    bs_mat <- data.table::dcast(scaled_bs, sample ~ target_id, 
+                                value.var = "scaled_reads_per_base")
+    # this now has the same format as the transcript matrix
+    # but it uses gene ids
+    bs_mat <- as.matrix(bs_mat[, -1])
+    rm(tidy_bs, scaled_bs)
+  }
+  
+  if (extra_bootstrap_summary) {
+    bs_quant_est_counts <- aperm(apply(bs_mat, 2,
```
Comment (pimentel) [inline R/bootstrap.R]: look into whether or not there is a faster matrixStats function. I think I tested the 1 that exists there and this is actually faster. will check soon

File: R/bootstrap.R [hunk 3]
```diff
@@ -359,9 +359,143 @@ dcast_bootstrap.kallisto <- function(obj, units, nsamples = NULL) {
...
+  
+  if (extra_bootstrap_summary) {
+    bs_quant_est_counts <- aperm(apply(bs_mat, 2, 
+                                       quantile))
+    colnames(bs_quant_est_counts) <- c("min", "lower", 
+                                       "mid", "upper", "max")
+    ret$bs_quants[[samp_name]]$est_counts <- bs_quant_est_counts
+  }
+  
+  bs_mat <- transform_fxn(bs_mat)
```
Comment (pimentel) [inline R/bootstrap.R]: change to `transform_fun` for consistency

**Comment by warrenmcg** [comment]
Don't worry about it at all!! Glad to see you back in action! :)

**Comment by pimentel** [comment]
merged in -- thanks again for all the hard work! will be making it into the next version scheduled for sometime next week!

**Pull request merged**